### PR TITLE
Streamline __call__ pattern between Tasks and Workflows

### PR DESCRIPTION
**Pull request by wild-endeavor**
Signed-off-by: wild-endeavor <wild-endeavor@users.noreply.github.com>

## TL;DR
Please read through the description in the [issue](https://github.com/flyteorg/flyte/issues/1032) as it's pretty succinct. 

Note that this PR only touches local execution.  Production runs on a real Flyte backend is not touched.

## Description

This PR basically further streamlines the logic in the `__call__` function of tasks and workflows.  At this point the logic is pretty much identical.  (But premature OOP inheritance pattern has bitten us in the past so we'll refrain from creating a base class out of it for now.)

## Unit Test Changes
We had to make some changes to the unit test to get this new paradigm to work.   You likely will need to make similar changes to your tests.

1.  Dataframe handling: If your task returned an instance of a Pandas dataframe as a `FlyteSchema` type in the past, since the Type Engine wasn't triggered, you just got the raw pandas dataframe.  Now you'll get the `FlyteSchema` object, so you'll need to change how you read that.  (See the change in `tests/flytekit/unit/core/test_type_hints.py`)
1. Sagemaker tasks: Probably nothing needs to happen but if you had modified sagemaker tasks to be tested locally, you'll need to provide real directories to the `train` and `validation` arguments.

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1032

### CODE REVIEW

File: flytekit/core/base_task.py [hunk 1]
```diff
@@ -286,16 +294,37 @@ def __call__(self, *args, **kwargs):
...
-                        mode=ExecutionState.Mode.LOCAL_TASK_EXECUTION, user_space_params=new_user_params
-                    )
+                    ctx.new_execution_state().with_params(mode=ExecutionState.Mode.LOCAL_WORKFLOW_EXECUTION)
                 )
-            ):
-                return self.execute(**kwargs)
+            ) as child_ctx:
+                result = self._local_execute(child_ctx, **kwargs)
+
+            expected_outputs = len(self.python_interface.outputs)
```
Comment (kumare3) [inline flytekit/core/base_task.py:305]: is the expected outputs check new?

File: flytekit/core/base_task.py [hunk 2]
```diff
@@ -286,16 +294,37 @@ def __call__(self, *args, **kwargs):
...
                 ctx.with_execution_state(
-                    ctx.execution_state.with_params(
-                        mode=ExecutionState.Mode.LOCAL_TASK_EXECUTION, user_space_params=new_user_params
-                    )
+                    ctx.new_execution_state().with_params(mode=ExecutionState.Mode.LOCAL_WORKFLOW_EXECUTION)
                 )
-            ):
-                return self.execute(**kwargs)
+            ) as child_ctx:
+                result = self._local_execute(child_ctx, **kwargs)
```
Comment (kumare3) [inline flytekit/core/base_task.py:303]: update  the comment in local execute. It is currently
```
        This code is used only in the case when we want to dispatch_execute with outputs from a previous node
        For regular execution, dispatch_execute is invoked directly.
```

I think it should be should be used when we have native values and want to execute?

**Comment by kumare3** [comment]
And if I did not mention - I think this makes a ton of sense - this makes all flyte transformations always required which is perfect

File: flytekit/core/base_task.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/core/base_task.py:305]: it is - copied from the workflow `__call__` function - just making things as similar as possible.

**Comment by wild-endeavor** [comment]
@jeevb thoughts?

**Review by kumare3 — DISMISSED** [review]

File: flytekit/core/promise.py [hunk 3]
```diff
@@ -793,7 +793,77 @@ def create_and_link_node(
...
     return create_task_output(node_outputs, interface)
+
+
+class LocallyExecutable(Protocol):
+    def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], Promise, VoidPromise]:
+        ...
+
+
+def executable_artifact_call_handler(entity: Union[SupportsNodeCreation, LocallyExecutable], *args, **kwargs):
+    """
```
Comment (kumare3) [inline flytekit/core/promise.py:808]: the logic is the same for workflows and tasks, but we did not want to create base class, hence a separate method

**Review by kumare3 — APPROVED** [review]
